daemon/gluster-blockd.c

```diff
@@ -43,6 +44,69 @@ static gbConfig *gbCfg;
 
 struct timeval TIMEOUT = {CLI_TIMEOUT_DEF, 0};  /* remote rpc call timeout, 5 mins */
 
+typedef struct globalData {
+  int    pfd;  /* pidfile file descriptor */
+  pid_t  chpid;
+} globalData;
+
+globalData ctx = {0,};
+
+
+void
+glusterBlockCleanGlobals(void)
+{
+  struct flock lock = {0, };
+
+    lock.l_type = F_UNLCK;
+    if (fcntl(ctx.pfd, F_SETLK, &lock) == -1) {
+      LOG("mgmt", GB_LOG_ERROR, "fcntl(UNLCK) on pidfile %s failed[%s]",
+          GB_LOCK_FILE, strerror(errno));
+    }
+
+    if (ctx.pfd) {
+      close(ctx.pfd);
+    }
+    glusterBlockDestroyConfig(gbCfg);
+}
+
+
+void onSigServerHandler(int signum)
+{
+  LOG("mgmt", GB_LOG_DEBUG,
+      "server process received signal:%d with pid:%lu", signum, getpid());
+
+  exit(EXIT_SUCCESS);  /* server process terminated by signal */
```

Comment:
Yeah, I was looking for something like that actually :-)

Was not aware (forget) there is a graceful way of exiting svc_run loop.

Thanks, let me test this and incorporate the change.

Reply:
Cool, before when debugging this issue I tested this code and it worked for me that time.
Thanks.

Reply:
@lxbsz 

BTW: where did you see svc_stop () ?

gluster-blockd.c:79:3: warning: implicit declaration of function ‘svc_stop’; did you mean ‘svc_stat’? [-Wimplicit-function-declaration]            
   svc_stop();

Reply:
May you mean svc_exit() ?

Reply:
Yeah, it s.
Sorry, I mixed the svc_exit with the other libs.

Reply:
For the libtirpc code, in svc_run() IMO, it should add the rdlock to protect the svc_pollfd, or it may cause the crash here.

Reply:
> For the libtirpc code, in svc_run() IMO, it should add the rdlock to protect the svc_pollfd, or it may cause the crash here.

Please ignore this too, currently the svc_exit() must  be called in the process's signal handler, or it will hit the crash issue.